src/control/ControlModule.cpp

```diff
@@ -0,0 +1,84 @@
+
+    void ControlModule::rotateRobotCommand() noexcept {
+        command.mutable_vel()->set_x(-command.vel().x());
+        command.mutable_vel()->set_y(-command.vel().y());
+        command.set_w(static_cast<float>(Angle(command.w() + M_PI)));
+    }
+
+    void ControlModule::limitRobotCommand() noexcept {
+        limitVel();
+        limitAngularVel();
+    }
+
+    void ControlModule::limitVel() noexcept {
+        auto limitedVel = Vector2(command.vel().x(), command.vel().y());
+
+        limitedVel = control::ControlUtils::velocityLimiter(limitedVel);
+
+        if (std::isnan(limitedVel.x) || std::isnan(limitedVel.y)) {
+            // TODO: Pass on the skill name to Control Module as well, so this error can be given correctly
+            //RTT_ERROR("Robot will have NAN: " + std::string{stp::getName()} + "!\nrobot: " +
+            //          std::to_string(robot.value()->getId()))
+        }
+
+        // Limit robot velocity when the robot has the ball
+        if (robot->hasBall() && limitedVel.length() > stp::control_constants::MAX_VEL_WHEN_HAS_BALL) {
+            // Clamp velocity
+            limitedVel = control::ControlUtils::velocityLimiter(limitedVel, stp::control_constants::MAX_VEL_WHEN_HAS_BALL,
+                                                                0.0, false);
+        }
+
+        command.mutable_vel()->set_x(static_cast<float>(limitedVel.x));
+        command.mutable_vel()->set_y(static_cast<float>(limitedVel.y));
+    }
+
+    void ControlModule::limitAngularVel() noexcept {
+        // Limit the angular velocity when the robot has the ball by setting the target angle in small steps
+        // Might want to limit on the robot itself
+        if (robot->hasBall() && command.use_angle()) {
+            auto targetAngle = command.w();
+            auto robotAngle = robot.value()->getAngle();
+
+            // If the angle error is larger than the desired angle rate, the angle command is adjusted
+            if (robotAngle.shortestAngleDiff(targetAngle) > stp::control_constants::ANGLE_RATE) {
+                // Direction of rotation is the shortest distance
+                int direction = Angle(robotAngle).rotateDirection(targetAngle) ? 1 : -1;
+                // Set the angle command to the current robot angle + the angle rate
+                command.set_w(static_cast<float>(robotAngle + Angle(direction * stp::control_constants::ANGLE_RATE)));
+            }
+        }
+    }
+
+    void ControlModule::publishRobotCommand(std::optional<::rtt::world::view::RobotView> robot_, const proto::RobotCommand& command_, const rtt::world::World *data) noexcept {
+        robot = robot_;
+        command = command_;
+
+        // If we are not left, commands should be rotated (because we play as right)
+        if (!SETTINGS.isLeft()) {
+            rotateRobotCommand();
+        }
+
+        limitRobotCommand();
+
+        io::io.publishRobotCommand(command, data);
```

Comment:
I don't like this line; the idea is that we collect all of the commands done over a tick, and then send them all at once; What I'd like to do is store these commands in e.g. a vector, and then have a different function which actually publishes them all at once (and checks if there are no double commands!)

Reply:
You don't have to worry about the publishing part if you don't want to, but if you can collect them in a vector that'd be nice.

Reply:
Ill first start with trying to store them in a vector, then see if i feel like trying the publishing :P